### PR TITLE
Add function to swap dataset longitude axis orientation

### LINKED ISSUE #135: Add option to specify longitude axis system when opening up a dataset

### PR DESCRIPTION
**Pull request by tomvothecoder**
## Description

- Closes #135 

This PR adds the ability to swap the axis orientation of longitude coordinates and bounds in the Dataset, either upon opening via `xcdat.open_dataset`/`xcdat.open_mfdataset` or as an independent function call. It also handles the prime meridian cell when converting from the 180 axis orientation to 360 by splitting it into east and west parts (reuses `_align_lon_bounds_to_360()` from the spatial averaging feature).

Summary of Changes
- Add `swap_lon_axis()` to `axis.py`
  - Supports axis orientations [-180, 180] and [0, 360]
  - Extendable to support subsets and wrapping beyond orientation min/max
- Add `lon_orient` kwarg to `open_dataset()` and `open_mfdataset()`  for specifying orientation upon opening a Dataset
  - Default behavior is to sort values in ascending order
- Add `_align_lon_to_360()` to handle prime meridian cell for the entire Dataset
- Move `_align_lon_bounds_to_360()` from `spatial_avg.py` to `axis.py`
  - Extract helper function `_get_prime_meridian_index()` from `_align_lon_bounds_to_360()`
  - Update `_get_longitude_weights()` to use `_get_prime_meridian_index()`

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

If applicable:

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass with my changes (locally and CI/CD build)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: tests/test_bounds.py [hunk 1]
```diff
@@ -63,7 +63,7 @@ def test_does_not_fill_bounds_for_coord_of_len_less_than_2(
     ):
         ds = self.ds_with_bnds.copy()
         ds = ds.isel(time=slice(0, 1))
-        ds = ds.drop("time_bnds")
+        ds = ds.drop_vars("time_bnds")
```
Comment (tomvothecoder) [inline tests/test_bounds.py:66]: To avoid a deprecation warning in the tests.

**Comment by tomvothecoder** [comment]
TODO @pochedls: add comments from meeting about covering some axis orientation edge cases and tag Jiwoo.

**Comment by pochedls** [comment]
> TODO @pochedls: add comments from meeting about covering some axis orientation edge cases and tag Jiwoo.

@tomvothecoder and @lee1043 - I've noted the CDAT behavior [here](https://github.com/XCDAT/xcdat/issues/135#issuecomment-966638180). This has some advantages. If you purely convert the bounds to 0-360 units, then a bound that spans the prime meridian has a range that may be unexpected (e.g., [-1, 1] becomes [359, 1]). This has created headaches for spatial averaging (since the weight in each dimension scales as the difference between the bounds; `1 - -1 = 2`, but `1 - 359 = -358`).

**Comment by tomvothecoder** [comment]
> If you purely convert the bounds to 0-360 units, then a bound that spans the prime meridian has a range that may be unexpected (e.g., [-1, 1] becomes [359, 1]). This has created headaches for spatial averaging (since the weight in each dimension scales as the difference between the bounds; 1 - -1 = 2, but 1 - 359 = -358).

We can pull some similar code from https://github.com/XCDAT/xcdat/pull/152 to handle this edge case at the Dataset level

**Comment by chengzhuzhang** [comment]
> We can pull some similar code from #152 to handle this edge case at the Dataset level
I was thinking the same, they are useful as general purpose functionality. :)

**Review by pochedls — COMMENTED** [review]
This is functionality that I have not used, but I suspect that users would want one change: a reordering of the longitude axis after the longitude units are converted. I *suspect* that a lot of what you would want to do might be complicated if you go from `[-180, -179, ..., -1, 0, 1, ..., 179]` to `[180, 181, ..., 359, 0, 1, ..., 179]` (e.g., plotting). There would ideally be a third step that also reorders the longitude axis (and the corresponding dataarrays and bounds) to `[0, 1, ..., 179, 180, 181, ..., 359]`.

Bounds are a special issue. The example above would have bounds of (notice the weird first bound):
```
lon_bnds = [[359.5, 0.5],
            [0.5, 1.5],
            ...,
            [178.5, 179.5],
            [179.5, 180.5],
            [180.5, 190.5], 
            ...,
            [358.5, 359.5]]
```
CDAT breaks this first bound up making the longitude axis go from length 360 to 361 (see added/modified values marked with asterisks): `[0, 1, ..., 179, 180, 181, ..., 359, *360*]` and
```
lon_bnds = [*[0, 0.5]*,
            [0.5, 1.5],
            ...,
            [178.5, 179.5],
            [179.5, 180.5],
            [180.5, 190.5], 
            ...,
            [358.5, 359.5], 
            *[359.5, 360.5]*]
```
This is what is done in https://github.com/XCDAT/xcdat/pull/152.

The corresponding data array (e.g., `tas`) would simply repeat the values for the first and last longitude. I think CDAT might actually modify the first and last longitude coordinate value, too (e.g., `[*0.5*, 1, ..., 179, 180, 181, ..., 359, *359.5*]`). 

@lee1043 - could you comment here on this PR (and let me know if you agree with this comment).

File: tests/test_axis.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                    dims=["lon"],
+                    attrs={"test_attr": "test"},
+                ),
+            },
+        )
+
+        assert result.identical(expected)
+
+    def test_swap_from_360_to_180_and_sorts(self):
+        # Dataset with longitude orientation 360
```
Comment (tomvothecoder) [inline tests/test_axis.py:91]:
```suggestion
```

File: xcdat/dataset.py [hunk 3]
```diff
@@ -43,6 +47,14 @@ def open_dataset(
         If True, decode times encoded in the standard NetCDF datetime format
         into datetime objects. Otherwise, leave them encoded as numbers.
         This keyword may not be supported by all the backends, by default True.
+    lon_orient: Optional[Literal[180, 360]], optional
+        If the Dataset has a longitude axis, use the specified orientation, by
+        default None.
```
Comment (tomvothecoder) [inline xcdat/dataset.py:52]:
```suggestion
    lon_orient: Optional[Literal[180, 360]], optional
        If the Dataset has a longitude axis, use the specified orientation. By
        default None.
```

File: xcdat/dataset.py [hunk 4]
```diff
@@ -143,6 +161,13 @@ def open_mfdataset(
...
```
Comment (tomvothecoder) [inline xcdat/dataset.py:166]:
```suggestion
        If the Dataset has a longitude axis, use the specified orientation. By
        default None.
```

**Comment by tomvothecoder** [comment]
Hey @pochedls and @lee1043, this PR is ready for review.

If you want to checkout this branch and test it, there are two ways:

1. Using `xcdat.open_dataset()`/`xcdat.open_mfdataset()` with kwarg `lon_orient` set to 180 or 360
    ```
    import xcdat
    
    ds = xcdat.open_dataset("path/to/file.nc", lon_orient=360)
    ```
2. Open dataset (can use xcdat or xarray) and pass object to `xcdat.swap_lon_axis()`
    ```
    import xcdat
    
    ds = xcdat.open_dataset("path/to/file.nc")
   # `sort_ascending` is optional
    ds = xcdat.swap_lon_axis(ds, to=360, sort_ascending=True)
    ```

**Comment by lee1043** [comment]
> Hey @pochedls this PR is ready for review. @lee1043 and @chengzhuzhang, if you want to test out this feature that'd be nice too.
> 
> After checking out the branch, there are two ways to test it:
> 
>     1. Using `xcdat.open_dataset()`/`xcdat.open_mfdataset()` with kwarg `lon_orient` set to 180 or 360
>        ```
>        import xcdat
>        
>        ds = xcdat.open_dataset("path/to/file.nc", lon_orient=360)
>        ```
> 
>     2. Open dataset (can use xcdat or xarray) and pass object to `xcdat.swap_lon_axis()`
>        ```
>        import xcdat
>        
>        ds = xcdat.open_dataset("path/to/file.nc")
>        # `sort_ascending` is optional
>        ds = xcdat.swap_lon_axis(ds, to=360, sort_ascending=True)
>        ```

@tomvothecoder it is great that this capability has been added, thanks! I like that there are 2 ways of using it. I have mostly used above way 1 in PMP, but have noticed that in some case I needed above way 2 to consider different areas with one open file (e.g., one area 60 to 120, another -30 to 30). 

Is there any chance that parameters for `lon_orient` or `to` of  `xcdat.swap_lon_axis()` to be more articulated? I am thinking of `[0, 360]` or `[-180, 180]`, instead of `360` or `180`, but open to any suggestion.

**Comment by pochedls** [comment]
I haven't reviewed this code yet, but the syntax looks fine for my use cases. 

@lee1043 is suggesting we should do a bit more to orient **and subset** in the same function call. An even more advanced version of this was suggested [here](([here](https://github.com/XCDAT/xcdat/discussions/169#discussion-3702033))) - it was noted we should even be able to "tile" the grid (e.g., -180 to 360 which would repeat part of the grid). 

@tomvothecoder - should this all be lumped together or should the subsetting be a future feature?

**Comment by lee1043** [comment]
> @lee1043 is suggesting we should do a bit more to orient **and subset** in the same function call. An even more advanced version of this was suggested here - it was noted we should even be able to "tile" the grid (e.g., -180 to 360 which would repeat part of the grid).

@pochedls thanks for your comment. I was not thinking that to parameter should be more flexible like [-180, 360] or any number e.g., [20, 270] or something. I was simply thinking `180` to [-180, 180] and `360` to [0, 360] so it is more straightforward. Although I like the idea of the flexibility, I think I am okay with having it for future feature.

File: xcdat/axis.py [hunk 5]
```diff
@@ -21,3 +24,256 @@
...
+    ----------
+    dataset : xr.Dataset
+         The Dataset containing a longitude axis.
+    to : Tuple[float, float]
+        The orientation to swap the Dataset's longitude axis to.
+
+        Supported orientations:
+
+          * (-180, 180): represents [-180, 180) in math notation
+          * (0, 360): represents [0, 360) in math notation
```
Comment (tomvothecoder) [inline xcdat/axis.py:47]: `to` kwarg is now a tuple of floats

File: xcdat/dataset.py [hunk 6]
```diff
@@ -43,6 +47,16 @@ def open_dataset(
...
         This keyword may not be supported by all the backends, by default True.
+    lon_orient: Optional[Tuple[float, float]], optional
+        The orientation to use for the Dataset's longitude axis (if it exists),
+        by default None.
+
+        Supported options:
+
+          * None:  use the current orientation (if the longitude axis exists)
+          * (-180, 180): represents [-180, 180) in math notation
+          * (0, 360): represents [0, 360) in math notation
```
Comment (tomvothecoder) [inline xcdat/dataset.py:58]: `lon_orient` kwarg is now a tuple of floats

**Review by jasonb5 — APPROVED** [review]
LGTM
